service/history/workflow/mutable_state_impl.go

```diff
@@ -4236,6 +4236,52 @@ func (ms *MutableStateImpl) RejectWorkflowExecutionUpdate(_ string, _ *updatepb.
 	return nil
 }
 
+func (ms *MutableStateImpl) AddWorkflowExecutionOptionsUpdatedEvent(
+	versioningOverride *workflowpb.VersioningOverride,
+) (*historypb.HistoryEvent, error) {
+	if err := ms.checkMutability(tag.WorkflowActionWorkflowOptionsUpdated); err != nil {
+		return nil, err
+	}
+	event := ms.hBuilder.AddWorkflowExecutionOptionsUpdatedEvent(versioningOverride)
+	if err := ms.ApplyWorkflowExecutionOptionsUpdatedEvent(event); err != nil {
+		return nil, err
+	}
+	return event, nil
+}
+
+func (ms *MutableStateImpl) ApplyWorkflowExecutionOptionsUpdatedEvent(event *historypb.HistoryEvent) error {
+	override := event.GetWorkflowExecutionOptionsUpdatedEventAttributes().GetVersioningOverride()
+	previousCurrentDeployment := ms.GetCurrentDeployment()
+	if override != nil {
+		ms.GetExecutionInfo().VersioningInfo.BehaviorOverride = override.GetBehavior()
+		ms.GetExecutionInfo().VersioningInfo.DeploymentOverride = override.GetDeployment()
+	} else {
+		ms.GetExecutionInfo().VersioningInfo.BehaviorOverride = enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED
+		ms.GetExecutionInfo().VersioningInfo.DeploymentOverride = nil
+	}
+	// If behavior was changed, deployment will also have changed, so we only need to check deployment
+	if !proto.Equal(ms.GetCurrentDeployment(), previousCurrentDeployment) {
```

Comment:
Do we need to compare effective behavior change too?

Reply:
yes, we do. Will do